### ISSUE TITLE
Vanilla Policy Gradient(VPG)

### ISSUE THREAD

**Issue by TypingCat** [TypingCat/wheel#27]
The main idea of [VPG](https://spinningup.openai.com/en/latest/algorithms/vpg.html) is to learn behaviors that cause good rewards. SPG(#14) uses rewards directly, but VPG uses advantage function. Discount rate is applied.

**Comment by TypingCat** [comment]
I analyzed the OpenAI VPG [code](https://github.com/openai/spinningup/blob/master/spinup/algos/pytorch/vpg/vpg.py). The following are new features that were not in SPG:

- Actor and critic networks
- Can choose between continuous and discrete action space
- Value: Reward that predicted by critic network
- Advantage: Discounted return

**Comment by TypingCat** [comment]
Generalized Advantage Estimation([GAE](https://arxiv.org/abs/1506.02438)) is tested on commit 2b7463d47b2af05d1200b89570a4ab1d2beba858. Instead to use single discount factor, GAE uses gamma and lambda. It helps to solve more complex control tasks. GAE will be applied to VPG.

**Comment by TypingCat** [comment]
I trained an actor-critic network for 1288 seconds with 60 agents. But it failed to learn the ability to move to the target. The following are parameters and results:

``` py
policy_std = torch.exp(torch.tensor([-0.5, -0.5]))
discount_gamma = 0.99
discount_lamda = 0.95
batch_size_max = 500
actor_learning_rate = 0.0001
critic_learning_rate = 0.001
critic_optimizer_iter = 80
```

``` bash
Batch: 507/500
[WARN] [wheel_navigation_vpg]: Time 13.0s, Actor Loss 0.1014, Critic Loss 0.0000
Batch: 501/500
[WARN] [wheel_navigation_vpg]: Time 20.8s, Actor Loss 0.1147, Critic Loss 0.0001
Batch: 530/500
[WARN] [wheel_navigation_vpg]: Time 24.8s, Actor Loss 0.0054, Critic Loss 0.0001
...
Batch: 4530/500
[WARN] [wheel_navigation_vpg]: Time 1060.8s, Actor Loss 1.2443, Critic Loss 0.0000
Batch: 566/500
[WARN] [wheel_navigation_vpg]: Time 1183.6s, Actor Loss 1.6278, Critic Loss 0.0000
Batch: 5937/500
[WARN] [wheel_navigation_vpg]: Time 1288.4s, Actor Loss 1.1907, Critic Loss 0.0000
```
